src/upload/index.ts

```diff
@@ -0,0 +1,201 @@
+      `${files.manifest}.gz`,
+    );
+    await this.uploadFile(gzipManifest, "application/gzip");
+    logger.info(`Upload: manifest file complete`);
+
+    await this.upload();
+  }
+
+  async createBlockFiles(
+    outputFileName: string,
+    previousSize: number,
+  ): Promise<BlockFile> {
+    this.deleteFileIfExists(outputFileName);
+    const manifestFileName = `${outputFileName}.manifest`;
+    this.deleteFileIfExists(manifestFileName);
+
+    let i = 1;
+    let currentByte = 0;
+    const outputFile = fs.createWriteStream(outputFileName);
+    const manifestFile = fs.createWriteStream(manifestFileName);
+
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      const block = await this.cache.getBlockBySequence(i);
+      if (block == null && previousSize === 0) break;
+      if (block == null) {
+        await this.waitForNextBlock();
+        continue;
+      }
+
+      const blockBuffer = LightBlock.encode(block).finish();
+      outputFile.write(blockBuffer);
+      manifestFile.write(
+        `${i},${currentByte},${currentByte + blockBuffer.byteLength - 1}\n`,
+      );
+      currentByte += blockBuffer.byteLength;
+
+      if (
+        !!previousSize &&
+        outputFile.bytesWritten >=
+          Math.max(this.chunkSizeMb * 1024 * 1024 + previousSize)
+      ) {
+        break;
+      }
+      i++;
+    }
+
+    outputFile.end();
+    manifestFile.end();
+
+    logger.info(
+      `New file upload created, size ${
+        outputFile.bytesWritten / 1024 / 1024
+      } MB, blocks: ${i - 1}`,
+    );
+    return { file: outputFileName, manifest: manifestFileName };
+  }
+
+  private deleteFileIfExists(fileName: string): void {
+    if (fs.existsSync(fileName)) {
+      fs.unlinkSync(fileName);
+    }
+  }
+
+  private async waitForNextBlock(): Promise<void> {
+    await new Promise((resolve) => setTimeout(resolve, 60000)); // Wait for 1 minute
+  }
+
+  async gzipFile(inputFile: string, outputFile: string): Promise<string> {
+    const writeStream = fs.createWriteStream(outputFile);
+    const readStream = fs.createReadStream(inputFile);
+    const gzip = zlib.createGzip();
+    readStream.pipe(gzip).pipe(writeStream);
+
+    await new Promise((resolve, reject) => {
+      writeStream.on("finish", resolve);
+      writeStream.on("error", reject);
+    });
+
+    readStream.close();
+    gzip.end();
+    writeStream.end();
+    logger.info(`Gzipping file complete: ${outputFile}`);
+    return outputFile;
+  }
+
+  async uploadFile(fileName: string, contentType: string): Promise<void> {
+    // due to consistentcy model of S3, should be safe to overwrite upload
+    const fileStream = fs.createReadStream(fileName);
+    const fileSize = fs.statSync(fileName).size;
+    const command = new PutObjectCommand({
+      Bucket: this.bucket,
+      ContentType: contentType,
+      ContentLength: fileSize,
+      Key: fileName,
+      Body: fileStream,
+    });
+    await this.s3Client.send(command);
```

Comment:
What is the retry logic used by default here (if any)?

Reply:
There is currently no retry logic. I suppose we could add an exponential backoff for this?